plugin/kotel/tracing.go

```diff
@@ -0,0 +1,162 @@
+	"go.opentelemetry.io/otel/propagation"
+	semconv "go.opentelemetry.io/otel/semconv/v1.12.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+var ( // interface checks to ensure we implement the hooks properly
+	_ kgo.HookProduceRecordBuffered   = new(Tracer)
+	_ kgo.HookProduceRecordUnbuffered = new(Tracer)
+	_ kgo.HookFetchRecordBuffered     = new(Tracer)
+)
+
+type Tracer struct {
+	tracerProvider trace.TracerProvider
+	propagators    propagation.TextMapPropagator
+	tracer         trace.Tracer
+}
+
+// TracingOption interface used for setting optional config properties.
+type TracingOption interface {
+	apply(*Tracer)
+}
+
+type tracingOptionFunc func(*Tracer)
+
+func (o tracingOptionFunc) apply(t *Tracer) {
+	o(t)
+}
+
+// NewTracer returns a Tracer, used as option for kotel to instrument franz-go with tracing
+func NewTracer(opts ...TracingOption) *Tracer {
+	t := &Tracer{}
+	for _, opt := range opts {
+		opt.apply(t)
+	}
+	if t.tracerProvider == nil {
+		t.tracerProvider = otel.GetTracerProvider()
+	}
+	if t.propagators == nil {
+		t.propagators = otel.GetTextMapPropagator()
+	}
+	t.tracer = t.tracerProvider.Tracer(
+		instrumentationName,
+		trace.WithInstrumentationVersion(SemVersion()),
+		trace.WithSchemaURL(semconv.SchemaURL),
+	)
+	return t
+}
+
+// TracerProvider takes a trace.TracerProvider and applies it to the Tracer
+// If none is specified, the global provider is used.
+func TracerProvider(provider trace.TracerProvider) TracingOption {
+	return tracingOptionFunc(func(t *Tracer) {
+		if t != nil {
+			t.tracerProvider = provider
+		}
+	})
+}
+
+// TracerPropagator takes a propagation.TextMapPropagator and applies it to the Tracer
+// If none is specified, the global Propagator is used.
+func TracerPropagator(propagator propagation.TextMapPropagator) TracingOption {
+	return tracingOptionFunc(func(t *Tracer) {
+		if t != nil {
+			t.propagators = propagator
+		}
+	})
+}
+
+// Hooks ---------------------------------------------------------------------
+
+// OnProduceRecordBuffered Begins the producer span
+func (t *Tracer) OnProduceRecordBuffered(r *kgo.Record) {
+	attrs := []attribute.KeyValue{
+		semconv.MessagingSystemKey.String("kafka"),
+		semconv.MessagingDestinationKindTopic,
+		semconv.MessagingDestinationKey.String(r.Topic),
+		semconv.MessagingKafkaClientIDKey.String(strconv.FormatInt(r.ProducerID, 10)),
+		semconv.MessagingOperationKey.String("send"),
+	}
+
+	opts := []trace.SpanStartOption{
+		trace.WithAttributes(attrs...),
+		trace.WithSpanKind(trace.SpanKindProducer),
+	}
+
+	spanContext, _ := t.tracer.Start(r.Context, fmt.Sprintf("%s send", r.Topic), opts...)
+
+	t.propagators.Inject(spanContext, NewRecordCarrier(r))
+
+	r.Context = spanContext
+}
+
+// OnProduceRecordUnbuffered Completes the producer span
+func (t *Tracer) OnProduceRecordUnbuffered(r *kgo.Record, err error) {
+	span := trace.SpanFromContext(r.Context)
+
+	span.SetAttributes(
+		semconv.MessagingMessageIDKey.String(strconv.FormatInt(r.Offset, 10)),
```

Comment:
Why no destination key here, but there is one above?

Reply:
Based on my understanding, since the span already exists and we are later referencing and closing it, the original attributes are inherited. This allows us to later add any additional data returned from Kafka to these existing attributes.